internal/langserver/handlers/command/module_list.go

```diff
@@ -0,0 +1,145 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/creachadair/jrpc2/code"
+	lsctx "github.com/hashicorp/terraform-ls/internal/context"
+	"github.com/hashicorp/terraform-ls/internal/langserver/cmd"
+	"github.com/hashicorp/terraform-ls/internal/terraform/datadir"
+	"github.com/hashicorp/terraform-ls/internal/uri"
+)
+
+const moduleListVersion = 0
+
+type moduleListResponse struct {
+	FormatVersion int               `json:"v"`
+	Modules       []terraformModule `json:"modules"`
+}
+
+type terraformModule struct {
+	Name             string            `json:"name"`
+	Path             string            `json:"path"`
+	Version          string            `json:"version"`
+	Type             string            `json:"type"`
+	DocLink          string            `json:"docklink"`
+	DependentModules []terraformModule `json:"depmodules"`
+}
+
+func ModuleListHandler(ctx context.Context, args cmd.CommandArgs) (interface{}, error) {
+	modUri, ok := args.GetString("uri")
+	if !ok || modUri == "" {
+		return nil, fmt.Errorf("%w: expected module uri argument to be set", code.InvalidParams.Err())
+	}
+
+	if !uri.IsURIValid(modUri) {
+		return nil, fmt.Errorf("URI %q is not valid", modUri)
+	}
+
+	modPath, err := uri.PathFromURI(modUri)
+	if err != nil {
+		return nil, err
+	}
+
+	mm, err := lsctx.ModuleFinder(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	found, _ := mm.ModuleByPath(modPath)
+	if found.ModManifest == nil {
+		return nil, nil
+	}
+
+	list := parseModuleRecords(found.ModManifest.Records)
+
+	return moduleListResponse{
+		Modules:       list,
+		FormatVersion: moduleListVersion,
+	}, nil
+}
+
+func parseModuleRecords(records []datadir.ModuleRecord) []terraformModule {
+	modules := make(map[string]terraformModule)
+	for _, manifest := range records {
+		if manifest.Dir == "." {
+			// this is the current directory, which is technically a module
+			// skipping as it's not relevant in the activity bar (yet?)
+			continue
+		}
+
+		moduleName := manifest.Key
+		subModuleName := ""
+
+		// determine if this module is nested in another module
+		// in the currecnt workspace by finding a period in the moduleName
+		// is it better to look at SourceAddr and compare?
+		if strings.Contains(manifest.Key, ".") {
+			v := strings.Split(manifest.Key, ".")
+			moduleName = v[0]
+			subModuleName = v[1]
+		}
+
+		// github modules are put in `files` directory, but Manifest Record does not
+		// record the `name` in Key it puts it in SourceAddr
```

Comment:
I think there's some misunderstanding as to what `Key` or `SourceAddr` are for.

`Key` is just locally-relevant reference name within the config (also known as the first and only label of the `module` block), `SourceAddr` is basically 1-1 copy of the raw `source` field within the `module` block.

For example:

```hcl
module "web_server" {
  source = "github.com/hashicorp/two-tier-infra"
}
```
this would be parsed as
 - `Key: "web_server"`
 - `SourceAddr: "github.com/hashicorp/two-tier-infra"`

Both are important and relevant, but not interchangeable.

Reply:
Thanks for the clarification, this makes much more sense.